File: 백준알고리즘/DFS(깊이 우선 탐색)/python/단지 번호 붙이기.py
import sys
input = sys.stdin.readline

N = int(input())
matrix = [list(input()) for _ in range(N)]

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]
cnt = 0
answer = []


def dfs(x, y):
    global cnt
    cnt += 1
    matrix[x][y] = '0'

    for i in range(4):
        nx = x + dx[i]
        ny = y + dy[i]

        if nx < 0 or nx >= N or ny < 0 or ny >= N:
            continue
        if matrix[nx][ny] == '1':
            dfs(nx, ny)


for i in range(N):
    for j in range(N):
        if matrix[i][j] == '1':
            cnt = 0
            dfs(i, j)
            answer.append(cnt)

print(len(answer))
answer.sort()
for item in answer:
    print(item)

File: 백준알고리즘/DFS(깊이 우선 탐색)/javascript/단지 번호 붙이기.js
const fs = require('fs');
const stdin = fs.readFileSync('/dev/stdin').toString().split('\n');

const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

N = +input();
const graph = [];
const visited = new Array(N).fill([]).map(() => new Array(N).fill(0));

const dx = [-1, 1, 0, 0];
const dy = [0, 0, -1, 1];

const answer = [];
var cnt = 0;

for (let i = 0; i < N; i++) {
  const a = input()
    .split('')
    .map(n => +n);

  graph.push(a);
}

function dfs(x, y) {
  cnt++;
  visited[x][y] = 1;

  for (let i = 0; i < 4; i++) {
    const nx = x + dx[i];
    const ny = y + dy[i];

    if (nx < 0 || nx >= N || ny < 0 || ny >= N) continue;

    if (graph[nx][ny] === 1 && visited[nx][ny] === 0) dfs(nx, ny);
  }
}

for (let i = 0; i < N; i++) {
  for (let j = 0; j < N; j++) {
    if (graph[i][j] === 1 && visited[i][j] === 0) {
      cnt = 0;
      dfs(i, j);
      answer.push(cnt);
    }
  }
}

answer.sort((a, b) => a - b);
console.log(answer.length);
for (let i = 0; i < answer.length; i++) {
  console.log(answer[i]);
}
